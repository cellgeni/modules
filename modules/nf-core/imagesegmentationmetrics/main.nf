process IMAGESEGMENTATIONMETRICS {
    label 'process_single'

    conda "${moduleDir}/environment.yml"

    container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
        'oras://community.wave.seqera.io/library/fire_monai_rasterio_tifffile_pruned:0957aceddbfc35bf' :
        'community.wave.seqera.io/library/fire_monai_rasterio_tifffile_pruned:106d8ac1dc78149c' }"

    input:
    tuple val(meta), path(ref_img), path(images2compare), val(metricname)

    output:
    tuple val(meta), path("*.csv"), emit: segmentation_metrics
    path "versions.yml"           , emit: versions

    when:
    task.ext.when == null || task.ext.when

    script:
    def args = task.ext.args ?: ''

    """
    export MPLCONFIGDIR=\$PWD/matplotlib_cache
    mkdir -p \$MPLCONFIGDIR

    segmentation_metrics.py run --path_ref ${ref_img} --path_list_seg "${images2compare}" --list_metrics "${metricname}" ${args}

    cat <<-END_VERSIONS > versions.yml
    ${task.process}:
        python: \$(python3 -c 'import platform; print(platform.python_version())')
        monai: \$(python3 -c 'import monai; print(monai.__version__)')
        seg-metrics: \$(segmentation_metrics.py version)

    END_VERSIONS
    """

    stub:
    def args = task.ext.args ?: ''
    def prefix = task.ext.prefix ?: "${meta}"
    // TODO nf-core: A stub section should mimic the execution of the original module as best as possible
    //               Have a look at the following examples:
    //               Simple example: https://github.com/nf-core/modules/blob/818474a292b4860ae8ff88e149fbcda68814114d/modules/nf-core/bcftools/annotate/main.nf#L47-L63
    //               Complex example: https://github.com/nf-core/modules/blob/818474a292b4860ae8ff88e149fbcda68814114d/modules/nf-core/bedtools/split/main.nf#L38-L54
    """
    touch ${prefix}.csv

    cat <<-END_VERSIONS > versions.yml
    ${task.process}:
        python: \$(python3 -c 'import platform; print(platform.python_version())')
        monai: \$(python3 -c 'import monai; print(monai.__version__)')
        seg-metrics: \$(segmentation_metrics.py version)

    END_VERSIONS
    """
}
