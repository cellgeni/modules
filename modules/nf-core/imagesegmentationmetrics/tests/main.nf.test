nextflow_process {

    name "Test Process IMAGESEGMENTATIONMETRICS"
    script "../main.nf"
    process "IMAGESEGMENTATIONMETRICS"

    tag "modules"
    tag "modules_nfcore"
    tag "imagesegmentationmetrics"

    test("imagesegmentationmetrics - csv") {

        when {
            process {
                """
                input[0] = [
                    "meta",
                    file('/lustre/scratch127/cellgen/cellgeni/yc6/Anno_UMAP/ref.tif', checkIfExists: true),
                    [file('/lustre/scratch127/cellgen/cellgeni/yc6/Anno_UMAP/label_removed_10.tif', checkIfExists: true),
                        file('/lustre/scratch127/cellgen/cellgeni/yc6/Anno_UMAP/label_removed_20.tif', checkIfExists: true),
                        file('/lustre/scratch127/cellgen/cellgeni/yc6/Anno_UMAP/label_removed_50.tif', checkIfExists: true)
                    ],
                    "DiceMetric MeanIoU GeneralizedDiceScore HausdorffDistanceMetric SurfaceDistanceMetric SurfaceDiceMetric MSEMetric MAEMetric RMSEMetric PSNRMetric"
                ]
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out).match() }
            )
        }

    }

    test("imagesegmentationmetrics - csv - stub") {

        options "-stub"

        when {
            process {
                """
                input[0] = [
                    "meta",
                    file('/lustre/scratch127/cellgen/cellgeni/yc6/Anno_UMAP/ref.tif', checkIfExists: true),
                    [file('/lustre/scratch127/cellgen/cellgeni/yc6/Anno_UMAP/label_removed_10.tif', checkIfExists: true),
                        file('/lustre/scratch127/cellgen/cellgeni/yc6/Anno_UMAP/label_removed_20.tif', checkIfExists: true),
                        file('/lustre/scratch127/cellgen/cellgeni/yc6/Anno_UMAP/label_removed_50.tif', checkIfExists: true)
                    ],
                    "DiceMetric MeanIoU GeneralizedDiceScore HausdorffDistanceMetric SurfaceDistanceMetric SurfaceDiceMetric MSEMetric MAEMetric RMSEMetric PSNRMetric"
                ]
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out).match() }
            )
        }

    }

}
