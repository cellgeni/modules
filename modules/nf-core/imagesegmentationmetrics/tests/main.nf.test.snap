{
    "imagesegmentationmetrics - csv - stub": {
        "content": [
            {
                "0": [
                    [
                        "meta",
                        "meta.csv:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "1": [
                    "versions.yml:md5,47b071734aad846571b08ea619d443f8"
                ],
                "segmentation_metrics": [
                    [
                        "meta",
                        "meta.csv:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,47b071734aad846571b08ea619d443f8"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.10.4"
        },
        "timestamp": "2025-03-25T14:03:42.361033"
    },
    "imagesegmentationmetrics - csv": {
        "content": [
            {
                "0": [
                    [
                        "meta",
                        "segmentation_benchmark_results.csv:md5,040aa0a7177e7056c68aaf8ca772e5b0"
                    ]
                ],
                "1": [
                    "versions.yml:md5,47b071734aad846571b08ea619d443f8"
                ],
                "segmentation_metrics": [
                    [
                        "meta",
                        "segmentation_benchmark_results.csv:md5,040aa0a7177e7056c68aaf8ca772e5b0"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,47b071734aad846571b08ea619d443f8"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.10.4"
        },
        "timestamp": "2025-03-25T14:03:25.427414"
    }
}